src/Swift.Runtime/src/Metadata/TypeMetadata.cs

```diff
@@ -0,0 +1,199 @@
+    /// <summary>
+    /// The metadata represents a struct
+    /// </summary>
+    Struct = 0 | TypeMetadataFlags.MetadataKindIsNonHeap,
+    /// <summary>
+    /// The metadata represents an enum
+    /// </summary>
+    Enum = 1 | TypeMetadataFlags.MetadataKindIsNonHeap,
+    /// <summary>
+    /// The metadata represents an optional type
+    /// </summary>
+    Optional = 2 | TypeMetadataFlags.MetadataKindIsNonHeap,
+    /// <summary>
+    /// The metadata represents an non-swift class
+    /// </summary>
+    ForeignClass = 3 | TypeMetadataFlags.MetadataKindIsNonHeap,
+    /// <summary>
+    /// The metadata represents an opaque type
+    /// </summary>
+    Opaque = 0 | TypeMetadataFlags.MetadataKindIsRuntimePrivate | TypeMetadataFlags.MetadataKindIsNonHeap,
+    /// <summary>
+    /// The metadata represents a tuple
+    /// </summary>
+    Tuple = 1 | TypeMetadataFlags.MetadataKindIsRuntimePrivate | TypeMetadataFlags.MetadataKindIsNonHeap,
+    /// <summary>
+    /// The metadata represents a closure/function
+    /// </summary>
+    Function = 2 | TypeMetadataFlags.MetadataKindIsRuntimePrivate | TypeMetadataFlags.MetadataKindIsNonHeap,
+    /// <summary>
+    /// The metadata represents a protocol
+    /// </summary>
+    Protocol = 3 | TypeMetadataFlags.MetadataKindIsRuntimePrivate | TypeMetadataFlags.MetadataKindIsNonHeap,
+    /// <summary>
+    /// The metadata represents a type of a TypeMetadata type
+    /// </summary>
+    Metatype = 4 | TypeMetadataFlags.MetadataKindIsRuntimePrivate | TypeMetadataFlags.MetadataKindIsNonHeap,
+    /// <summary>
+    /// The metadata represents an Objective C wrapper
+    /// </summary>
+    ObjCClassWrapper = 5 | TypeMetadataFlags.MetadataKindIsRuntimePrivate | TypeMetadataFlags.MetadataKindIsNonHeap,
+    /// <summary>
+    /// The metadata represents a type of an existential container
+    /// </summary>
+    ExistentialMetatype = 6 | TypeMetadataFlags.MetadataKindIsRuntimePrivate | TypeMetadataFlags.MetadataKindIsNonHeap,
+    /// <summary>
+    /// The metadata represents a heap local variable
+    /// </summary>
+    HeapLocalVariable = 0 | TypeMetadataFlags.MetadataKindIsNonType,
+    /// <summary>
+    /// The metadata represents a generic heap local variable
+    /// </summary>
+    HeapGenericLocalVariable = 0 | TypeMetadataFlags.MetadataKindIsNonType | TypeMetadataFlags.MetadataKindIsRuntimePrivate,
+    /// <summary>
+    /// The metadata represents an error
+    /// </summary>
+    ErrorObject = 1 | TypeMetadataFlags.MetadataKindIsNonType | TypeMetadataFlags.MetadataKindIsRuntimePrivate,
+    // Swift source code says that for fixed values, this will never exceed 0x7ff,
+    // but all class types will be 0x800 and above
+    /// <summary>
+    /// The metadata represents a class
+    /// </summary>
+    Class = 0x800 
+}
+
+
+/// <summary>
+/// Represents the type metadata for a Swift type
+/// </summary>
+public readonly struct TypeMetadata : IEquatable<TypeMetadata> {
```

Comment:
How this correlates with `SwiftMetadata`? Do we plan to use metadata for type semantics on the C# side or only as an unmanaged pointer for callconv?

Reply:
SwiftMetadata is based off the same type, but the difference here is that I'm not trying to define it as a ref struct but an struct with an opaque pointer so that it's easy to manipulate, compare, hash, and pass to and from p/invokes (the most common operations by far) and because it is going to escape the stack.
SwiftMetadata as is feels like a binding-time type not a runtime type.
I learned early on that we can't rely on it at binding time because of generics. It appears the only thing it gets used for at binding time is putting in a `Size` value in the `StructLayout` attribute.
For this I have several questions: 1 - do we need that? 2 - if we do, can't we get it in some other way (for example, follow the steps in the doc'n [here](https://github.com/swiftlang/swift/blob/main/docs/ABI/TypeLayout.rst#fragile-struct-and-tuple-layout).
In the larger picture, we need a binding-time type database, which is it's own thing, but certainly one of the things we can put into in optional properties for Size, Stride, and Alignment so that we can calculate a new Struct's size.  But like I said, that's a larger discussion.

Reply:
> 1 - do we need that?
I think this might be relevant https://github.com/dotnet/runtimelab/pull/2796#discussion_r1852658458. tldr: we need to check

Reply:
Did not mean generics, but having an explicit size value in the `StructLayout` attribute for frozen structs. @matouskozak shared a great example https://github.com/dotnet/runtimelab/pull/2796#discussion_r1854188388